File: Data Structures & Algorithms/two-integer-sum/submission-16.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[][] A = new int[nums.length][2];
        for (int i = 0; i < nums.length; i++) {
            A[i][0] = nums[i];
            A[i][1] = i;
        }

        Arrays.sort(A, Comparator.comparingInt(a -> a[0]));

        int l = 0;
        int r = A.length - 1;

        while (l < r) {
            int curr = A[l][0] + A[r][0];

            if (curr == target) {
                int idx1 = Math.min(A[l][1], A[r][1]);
                int idx2 = Math.max(A[l][1], A[r][1]);
                return new int[]{idx1, idx2};

            } else if (curr < target) {
                l++;

            } else {
                r--;
            }
        }

    return new int[]{-1, -1};
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-16.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [[0] * 2 for _ in range(n + 1)] 
        dp[n][0] = -1e6
        dp[n][1] = 0

        for i in range(n - 1, -1, -1):
            dp[i][1] = max(0, nums[i] + dp[i + 1][1])
            dp[i][0] = max(dp[i + 1][0], nums[i] + dp[i + 1][1])
        
        return dp[0][0]
        
        def dfs(i, started):
            if i == len(nums):
                return 0 if started else -1e6

            if started:
                return max(0, nums[i] + dfs(i + 1, True))
            
            return max(dfs(i + 1, False), nums[i] + dfs(i + 1, True))
        
        return dfs(0, False)

    